custom/rch/views.py

```diff
@@ -0,0 +1,335 @@
+        Location Hierarchy: State -> District -> Village -> AWC
+        Filters down-up AWC -> Village -> District -> State
+        Filters by AWC id present as owner in cas records.
+        Tries to filters by AWC id selected or AWCs under village ID selected but then falls back to
+        district's or state's AWC ids for filtering
+        When filtering using Village id, it looks for the corresponding AWC ids using
+        AWC-Village mapping but it would result in beneficiaries from other Village since
+        its many to many mapping.
+        :param include_matched_beneficiaries: to include matched beneficiaries or
+         just render rch records excluding the ones that have been matched
+        :return: True if it could find records else False
+        """
+        self.beneficiaries = case_es.CaseES().domain(ICDS_CAS_DOMAIN).size(RECORDS_PER_PAGE)
+        if not include_matched_beneficiaries:
+            self.beneficiaries = self.beneficiaries.filter(case_es.missing('rch_record_id'))
+        awc_ids = []
+
+        # find corresponding awc_id(s) for the selected village id to search
+        # cas records by their awc_ids set as owner_id
+        village_code = in_data.get('village_code')
+        if village_code:
+            awc_ids = awc_ids + AreaMapping.fetch_awc_ids_for_village_id(village_code)
+
+        awc_id = in_data.get('awcid')
+        if awc_id:
+            awc_ids = awc_ids + [awc_id]
+
+        # if no village or awc then simply filter by all AWC IDs under the district
+        if not awc_ids:
+            dtcode = in_data.get('dtcode')
+            if dtcode:
+                awc_ids = AreaMapping.fetch_awc_ids_for_district(dtcode)
+
+        # if no village or awc then simply filter by all AWC IDs under the state
+        if not awc_ids:
+            stcode = in_data.get('stcode')
+            if stcode:
+                awc_ids = AreaMapping.fetch_awc_ids_for_state(stcode)
+
+        if awc_ids:
+            awc_loc_ids = list(
+                SQLLocation.objects.filter(site_code__in=awc_ids).values_list('location_id', flat=True)
+            )
+            user_ids = user_ids_at_locations(awc_loc_ids)
+            # ToDo: Confirm that this check is correct to include both users and locations
+            # ToDo: remove awc_ids added for demo purpose
+            user_ids = user_ids + awc_loc_ids + awc_ids
+            if user_ids:
+                self.beneficiaries = self.beneficiaries.filter(case_es.user(user_ids))
+                return True
+
+        # set beneficiaries empty if unable to filter by any awc_id
+        self.beneficiaries = []
+        return False
+
+    def _set_context_for_cas_records(self, context, in_data=None):
+        include_matched_beneficiaries = (in_data.get('matched') == 'on')
+        size = in_data.get('limit')
+        offset = (in_data.get('page') - 1) * in_data.get('limit')
+        if self._set_cas_beneficiaries(include_matched_beneficiaries, in_data=in_data):
```

Comment:
this should rather just return the beneficiaries rather than setting them as an attribute of the class:
```
beneficiaries = self._get_cas_beneficiaries(include_matched_beneficiaries, in_data=in_data)
if beneficiaries:
   ....
```